### PR TITLE
fix: Added dismiss button to snackbar messages

### LINKED ISSUE #6218: Add dismiss button to snackbar messages
Currently, our snackbar messages look like this:

Screenshot 2022-03-15 at 08 24 49

While the message and CTA are pretty clear, I find that there is no obvious way to dismiss them. We currently support clicking anywhere on the snackbar to remove them, but there are two issues with this approach:

1. Not everyone knows this (I didn't!), since there is no visible affordance that indicates it. 
2. Because focus cannot be placed on the snackbar itself, they can only be dismissed if you are using a mouse or touch input device. Keyboard-only users have not much of a choice.

Snackbars are also intended to [disappear automatically](https://material.io/components/snackbars#behavior):

> Snackbars appear without warning, and don't require user interaction. They automatically disappear from the screen after a minimum of four seconds, and a maximum of ten seconds.

But in our case, it stays in view until clicked on.

I don't mind having them stay in view, it makes sense given the type of messages we are using them for. But I think **we can improve the interaction if we add a close/dismiss button to them**.

This is similar to what GMail is doing:

Captura de Pantalla 2022-03-16 a la(s) 10 45 26 a m

For us, it could look like this:

Captura de Pantalla 2022-03-16 a la(s) 10 45 26 a m

- The puzzle piece emoji (or any emoji we may use in the future) is relocated to the start of the text. This anchors the eyes, but also brings visual balance to the whole component.
- The `Learn more` button maintains its primary CTA status by visually standing out more than anything else.
- The close button is visible, but does not steal any attention from the message nor the primary CTA.

### Acceptance criteria
- Users can dismiss a snackbar via a visible affordance (dimiss button)

### Estimated effort
(Needs input from a developer)

### PR DESCRIPTION
**Pull request by 87prashant**
# Description

fixes #6218

## Demo

Screenshot 2022-06-22 140132

Screenshot 2022-06-22 140046

## Testing scenarios

Try to generate different snackbar messages and verify the dismiss button and its functionality.

### CODE REVIEW

File: packages/client/components/SnackbarMessageAction.tsx [hunk 1]
```diff
@@ -3,23 +3,20 @@ import {SnackAction} from './Snackbar'
...
   color: PALETTE.ROSE_500,
   fontSize: 14,
   fontWeight: 600,
   marginLeft: 8,
   padding: 8,
-  transition: `background 100ms ${DECELERATE}`,
-  ':hover,:focus,:active': {
-    backgroundColor: '#ffffff0e'
-  }
+  backgroundColor: '#ffffff0e'
```
Comment (igorlesnenko) [inline packages/client/components/SnackbarMessageAction.tsx:19]: -1 I think the action button still need to react when I hover my cursor over it.

https://user-images.githubusercontent.com/466991/177758630-173c497e-1d5e-434e-9e58-ff80b1642c5e.mov

cc @enriquesanchez

**Review by igorlesnenko — CHANGES_REQUESTED** [review]
Thank you for taking this issue! A few changes need to be made

File: packages/client/components/SnackbarMessageAction.tsx (hunk 1 above)
Comment (87prashant) [reply packages/client/components/SnackbarMessageAction.tsx:19]: OK

File: packages/client/components/SnackbarMessage.tsx [hunk 2]
```diff
@@ -71,13 +89,15 @@ const SnackbarMessage = (props: Props) => {
...
-        onClick={dismissSnack}
         onMouseEnter={onMouseEnter}
         onMouseLeave={onMouseLeave}
       >
         <Text>{message}</Text>
         <SnackbarMessageAction action={action} />
         <SnackbarMessageAction action={secondaryAction} />
+        <DismissButton onClick={dismissSnack}>
+          <StyledIcon>close</StyledIcon>
+        </DismissButton>
```
[the remark this answers is not in the document]
Comment (87prashant) [reply packages/client/components/SnackbarMessage.tsx:99]: Thanks for helping @igorlesnenko 
I found  `noDismissOnClick: true`
I think noDismissOnClick should be that element.
Am I right?

[the remark this answers is not in the document]
Comment (igorlesnenko) [reply packages/client/components/SnackbarMessage.tsx:99]: @87prashant

I would follow this recommendation from @ackernaut  https://github.com/ParabolInc/parabol/issues/6218#issuecomment-1163543678:

>  I would not recommend touching every existing snack bar in the code base as a refactor. I think if this pattern is available to us we can add it in as the need arises

So I suggest to not modify any of existing snackbars, but just add an option to enable this close button. Then if needed we can create a new issue and define which of the current snackbars needs to have this close button.

[the remark this answers is not in the document]
Comment (87prashant) [reply packages/client/components/SnackbarMessage.tsx:99]: Ok @igorlesnenko
But I think we need the previous behavior of the snackbar (i.e. click anywhere to dismiss) as we are not providing any dismiss button.

[the remark this answers is not in the document]
Comment (igorlesnenko) [reply packages/client/components/SnackbarMessage.tsx]: @87prashant yes, we keep all previous behaviour, but just have a new ability to display close button.

File: packages/client/components/SnackbarMessage.tsx [hunk 3]
```diff
@@ -88,15 +88,24 @@ const SnackbarMessage = (props: Props) => {
     onMouseLeave
   } = props
   useTransitionEnd(300, status, onTransitionEnd)
+  //Currently do not need dismiss button
+  const displayDismissButton = false
```
Comment (igorlesnenko) [inline packages/client/components/SnackbarMessage.tsx:92]: It is now hardcoded for all snackbars. What was meant is that we want to be able to enable this close button individually per snackbar.

So can we make this option to be possible to specify like this: 

```
atmosphere.eventEmitter.emit('addSnackbar', {
        autoDismiss: 5,
        message: 'Hello world,
        showDismissButton: true,
        
``` 

?

**Review by igorlesnenko — CHANGES_REQUESTED** [review]

File: packages/client/components/SnackbarMessage.tsx (hunk 3 above)
Comment (87prashant) [reply packages/client/components/SnackbarMessage.tsx:92]: Ok

Comment (87prashant) [reply packages/client/components/SnackbarMessage.tsx:92]: I hope I covered all snackbars.

**Review by igorlesnenko — CHANGES_REQUESTED** [review]

File: packages/client/components/SnackbarMessage.tsx [hunk 4]
```diff
@@ -64,20 +86,26 @@ const SnackbarMessage = (props: Props) => {
...
   } = props
   useTransitionEnd(300, status, onTransitionEnd)
   return (
     <Space>
       <MessageStyles
         status={status}
-        onClick={dismissSnack}
         onMouseEnter={onMouseEnter}
         onMouseLeave={onMouseLeave}
+        onClick={showDismissButton ? undefined : dismissSnack}
```
[the remark this answers is not in the document]
Comment (87prashant) [reply packages/client/components/SnackbarMessage.tsx:99]: Ok

**Review by igorlesnenko — CHANGES_REQUESTED** [review]
Just a couple of things left

1) Let's bring back the bg color around action button, and update hover color. It will look like this:

```
const Action = styled(PlainButton)({
  borderRadius: Radius.SNACKBAR,
  color: PALETTE.ROSE_500,
  fontSize: 14,
  fontWeight: 600,
  marginLeft: 8,
  padding: 8,
  backgroundColor: '#ffffff17',
  transition: `background 100ms ${DECELERATE}`,
  ':hover,:focus,:active': {
    backgroundColor: '#ffffff26'
  }
})
```

2)  Specifying showDismissButton works, but typescript shows an error. 

image

**Comment by 87prashant** [comment]
> Specifying showDismissButton works, but typescript shows an error.

Errors are not generating on my system. If remember correctly these errors went away automatically after some time.

Screenshot 2022-07-18 163135

**Comment by igorlesnenko** [comment]
@87prashant you are right!  Looks like I had some IDE cache

**Review by igorlesnenko — APPROVED** [review]
Looks good to me!

**Comment by mattkrick** [comment]
Great job @87prashant!

I really enjoyed reading through this issue & seeing how it evolved from a mandatory button to something message-specific. Great minimal UX!

**Pull request merged**